pkg/testing/ess/serverless_provision.go

```diff
@@ -71,114 +62,118 @@ func (prov *ServerlessProvision) SetLogger(l runner.Logger) {
+	srvReq := ServerlessRequest{Name: request.ID, RegionID: prov.cfg.Region}
 
-			}
-			depUp <- isUp
-		}()
+	prov.log.Logf("Creating serverless stack %s [stack_id: %s]", request.Version, request.ID)
+	proj, err := client.DeployStack(createCtx, srvReq)
+	if err != nil {
+		return runner.Stack{}, fmt.Errorf("error deploying stack for request %s: %w", request.ID, err)
+	}
+	err = client.WaitForEndpoints(createCtx)
+	if err != nil {
+		return runner.Stack{}, fmt.Errorf("error waiting for endpoints to become available for request: %w", err)
+	}
+	stack := runner.Stack{
+		ID:            request.ID,
+		Version:       request.Version,
+		Elasticsearch: client.proj.Endpoints.Elasticsearch,
+		Kibana:        client.proj.Endpoints.Kibana,
+		Username:      client.proj.Credentials.Username,
+		Password:      client.proj.Credentials.Password,
+		Internal: map[string]interface{}{
+			"deployment_id":   proj.ID,
+			"deployment_type": proj.Type,
+		},
+		Ready: false,
 	}
+	prov.log.Logf("Created serverless stack %s [stack_id: %s, deployment_id: %s]", request.Version, request.ID, proj.ID)
+	return stack, nil
+}
+
+// WaitForReady should block until the stack is ready or the context is cancelled.
+func (prov *ServerlessProvision) WaitForReady(ctx context.Context, stack runner.Stack) (runner.Stack, error) {
+	deploymentID, deploymentType, err := prov.getDeploymentInfo(stack)
+	if err != nil {
+		return stack, fmt.Errorf("failed to get deployment info from the stack: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
+	defer cancel()
+
+	client := NewServerlessClient(prov.cfg.Region, "observability", prov.cfg.APIKey, prov.log)
+	client.proj.ID = deploymentID
+	client.proj.Type = deploymentType
+	client.proj.Region = prov.cfg.Region
+	client.proj.Endpoints.Elasticsearch = stack.Elasticsearch
+	client.proj.Endpoints.Kibana = stack.Kibana
+	client.proj.Credentials.Username = stack.Username
+	client.proj.Credentials.Password = stack.Password
+
+	prov.log.Logf("Waiting for serverless stack %s to be ready [stack_id: %s, deployment_id: %s]", stack.Version, stack.ID, deploymentID)
+
+	errCh := make(chan error)
+	var lastErr error
+
+	ticker := time.NewTicker(30 * time.Second)
+	defer ticker.Stop()
 
-	gotUp := 0
 	for {
 		select {
 		case <-ctx.Done():
-			return nil, ctx.Err()
-		case err := <-depErrs:
-			return nil, fmt.Errorf("error waiting for stacks to become available: %w", err)
-		case isUp := <-depUp:
-			if isUp {
-				gotUp++
+			if lastErr == nil {
+				lastErr = ctx.Err()
 			}
-			if gotUp >= len(requests) {
-				return stacks, nil
+			return stack, fmt.Errorf("serverless stack %s [stack_id: %s, deployment_id: %s] never became ready: %w", stack.Version, stack.ID, deploymentID, lastErr)
+		case <-ticker.C:
+			go func() {
+				statusCtx, statusCancel := context.WithTimeout(ctx, 30*time.Second)
+				defer statusCancel()
+				ready, err := client.DeploymentIsReady(statusCtx)
```

Comment:
It seems like the stateful `DeploymentIsReady` method does this waiting work, and `WaitForReady` just wraps it, where as this `DeploymentIsReady` just gives you the current status.

If you call `WaitForReady` this gives you the same result for both, but calling `DeploymentIsReady` doesn't. This might bite someone in the future since it isn't consistent.

Reply:
I agree it is not consistent. This PR is about changing to use CFT not about making cloud and serverless consistency between the implementations, which I agree should be fixed (but not in this PR). This PR is about switching to the CFT region and making it keep state when a stack doesn't become ready. This has now snowballed into a review of serverless implementation, which would have been better to bring up when it was originally added.

Reply:
It wasn't clear that it was already like this before, thanks.